### DISCUSSION TITLE
Trigger `watch` by using `setValue` as soon as component is rendered

### DISCUSSION THREAD

**Discussion by giovannipiller** [react-hook-form/react-hook-form#6054] [Q&A]
Follow-up to #6034
https://github.com/react-hook-form/react-hook-form/issues/6034#issuecomment-884908150

-------

> use setValue will trigger watch to update input change.

I've been trying to implement this solution by removing the `defaultValue` from the `useController(...)`, and setting it using a `useEffect` when the component is rendered. Something like this:

```typescript
const myOwnDefaultValue = true;
React.useEffect(() => {
  // checking for undefined, to simulate setting a default value
  if (field.value === undefined)
    setValue("hasUserAcceptedPrivacyPolicy", myOwnDefaultValue);
}, [field.value, setValue]);
```

However, while the value for `hasUserAcceptedPrivacyPolicy` is effectively set to `true`, the `useWatch` won't trigger. Therefore, `<UserEmail>` won't display an input.

https://codesandbox.io/s/wizardly-pasteur-rq8z1?file=/src/App.tsx

## Hacky workaround

I noticed that letting the `useEffect` run twice, makes things work. I'm assuming that it's just a coincidence, probably something related to `watch` beginning to subscribe to changes only after a some other event.

 ```typescript
const myOwnDefaultValue = true;
React.useEffect(() => { 
  if (field.value === undefined || field.value === myOwnDefaultValue)
    setValue("hasUserAcceptedPrivacyPolicy", myOwnDefaultValue);
}, [field.value, setValue]);
```

https://codesandbox.io/s/clever-hooks-whzud?file=/src/App.tsx

----

Another _terrible_ hack would be using a `setTimeout` or some sort of empty `async/await`, to delay the run of the first `useEffect`.
This again makes me think that I'm calling the `setValue` way before RHF has started watching for value changes to notify via `watch`.

```typescript
const myOwnDefaultValue = true;
React.useEffect(() => {
  setTimeout(() => {
    if (field.value === undefined)
      setValue("hasUserAcceptedPrivacyPolicy", myOwnDefaultValue);
  });
}, [field.value, setValue]);
```

This of course doesn't look like something sane to do 😰.

https://codesandbox.io/s/brave-margulis-70ic1?file=/src/App.tsx

----

Am I doing something wrong here?

**Comment by giovannipiller** [comment]
@bluebill1049 I've opened a discussion here, as we talked in #6034

just pinging you since it probably got lost between all the other threads 🕵️‍♂️

**Reply by bluebill1049** [reply]
why not set it at the form level?  `setValue` and `reset` both works.

https://codesandbox.io/s/admiring-bhabha-88q3s

**Reply by giovannipiller** [reply]
> why not set it at the form level? `setValue` and `reset` both works.

Sorry, this detail got lost between the various discussions.
In my scenario `UserEmail` is the component tasked with either retrieving or generating the `defaultValue`.

A couple of other examples:
1) `UserEmail` could perform an AJAX to fetch the email address to be used as `defaultValue`
2) a component similar to `UserEmail`, where it accepts a prop to customize `name`.

In both these scenarios, it would be very hard to `setValue` at form level.

----

To expand on my use-case: I'm creating components with `useController` to build forms that can be as modular as possible.
Using `UserEmail` as an example: should I need to include a user email field in a new form, I could just reuse the same component everywhere.

**Reply by bluebill1049** [reply]
>You need to either set defaultValue at the field-level or call useForm with defaultValues. If both are set,devaultValues will be used.

<img width="953" alt="Screen Shot 2021-07-26 at 6 15 44 pm" src="https://user-images.githubusercontent.com/10513364/126956349-378877ab-aa21-40b4-8bfe-b0b7720dfb04.png">

- you will need to provide either `defaultValue` at useController or `defaultValues` at `useForm`

I wouldn't set up input value inside the `useController` or `Controller`, because their job is only for collecting value and validate the inputs.

perhaps you want to update the input value at your UserEmail component such as below?

https://codesandbox.io/s/admiring-bhabha-88q3s?file=/src/App.tsx

**Reply by giovannipiller** [reply]
But what if I want to keep both `AcceptPrivacyPolicy` and `UserEmail` completely separate from the form?
In other words: I don't want the `useForm` to know anything about their existence.

----

To provide further context: I have an application which builds forms based on many external factors (users can add new inputs and implement cross-field relationship that are not that different from `AcceptPrivacyPolicy` and `UserEmail`).
My idea was to keep the `useForm` as clean as possible, with just enough code to setup the form.
Then, based on various conditions, dynamically include components that would set up `useController` with their own default values and validations.

Due to the complexity of my own code, which includes plenty of stuff that is not RHF-related, I opted for creating the simplified example that we're discussing now (`AcceptPrivacyPolicy` and `UserEmail`).

What I'm looking for, is a way to create a relationship between those two fields without `useForm` being involved at all.
Which is why I initially thought that it was wrong for `useWatch` to not trigger whenever a defaultValue is added to the form data (#6034).

I can provide more examples, but wanted to keep the complexity low.

Thanks again for your help.

**Comment by giovannipiller** [comment]
@bluebill1049 
I didn't mean to disturb you this much, but I still think it's worthy of a discussion 😰 Sorry

I have been trying to find a proper way to implement this scenario, but as far as I can tell there's very little I can do outside my [hacky](https://codesandbox.io/s/clever-hooks-whzud?file=/src/App.tsx) [workarounds](https://codesandbox.io/s/brave-margulis-70ic1?file=/src/App.tsx).

## To recap:
1. `<AcceptPrivacyPolicy>` uses `useController` to set up a field
   - the field starting value it's not known to the `useForm` (in real life could be determined by an API call, or something else that only `<AcceptPrivacyPolicy>` knows about)
   - in my examples, I tried setting this starting value from `<AcceptPrivacyPolicy>` by:
     1. using `useController({ defaultValue: .... })` – [Codesandbox](https://codesandbox.io/s/distracted-grass-j89ve?file=/src/App.tsx)
     2. using `setValue(...)` after setting up the `useController` – which didn't work unless I used my previously mentioned [hacky](https://codesandbox.io/s/clever-hooks-whzud?file=/src/App.tsx) [workarounds](https://codesandbox.io/s/brave-margulis-70ic1?file=/src/App.tsx).
2.  `<UserEmail>` uses `useWatch` to set up a dependency on the field managed by `<AcceptPrivacyPolicy>`
3. in my scenario, the component that sets  up `useForm(...)` has no way of knowing **if or when** those components will be included. Because of this, I can't simply use it to provide a `defaultValues` to be watched by its child components.


### Expected behavior
1. `<AcceptPrivacyPolicy>` is added to the form and sets its starting value.
2. `<UserEmail>`'s `useWatch` reacts to this value by adding something to the page (in this scenario, a new input).

### Actual behavior
1. `<AcceptPrivacyPolicy>` is added to the form and sets its starting value.
2. `<UserEmail>`'s `useWatch` will never react to this value as it's added to the form.

Outside of my hacks, I have not been able to setup a proper implementation of this scenario.
`<UserEmail>` has no real way of knowing that the property `hasUserAcceptedPrivacyPolicy` has been added to the form data.

## Trying to simplify even further
If I have a component `<A>` that reacts to the value of a field with name `field1`. The fact that a value for `field1` is set via `defaultValue` of a `useController` is of extreme importance to `<A>`. There's a huge difference in a form between:
1. not having a value for `field1`, or 
2. having that value being its `defaultValue`

And the component `<A>` would like to respond to that difference.

## A case for watching `defaultValue`

Now, I know that `useWatch` is not supposed to react to `defaultValues`, but I still believe that there should be a way to properly subscribe and react to those values being added to the form data.

This is especially important due to the ability of `useController` to change a form's data by providing its own `defaultValue`.

I can think of multiple scenarios where I can't reliably provide the entire structure of `defaultValues` right away to a `useForm`.
And since components with `useController` can modify the form data with their `defaultValue`, there should be a way for _other_ components to track those changes.

Or at the very least, there should be a way to use something like `setValue`, in place of `defaultValue`, to let components that are subscribed via `useWatch` that a value has been set. Something that I have not been able to do [without a hack](https://codesandbox.io/s/clever-hooks-whzud?file=/src/App.tsx).

----

I know it's quite a wall of text, but I'm just really passionate about react-hook-form ❤️
It's the first form library that I actually enjoy using, and it frustrates me to no end that I can't seem to solve this problem.

**Reply by bluebill1049** [reply]
The problem is related to the understanding of the react component execution order. 

1. react will start with the child component
2. the component order does matter

- AcceptPrivacyPolicy (you are setting the value in here)
- UserEmail (but your subscription starts here, at useEffect of useWatch, so the setValue is only updated the hook form's internal value, and won't send the signal to the `useWatch`)

You will see that if you switch the order of your component.

------------

Also, you should provide either defaultValues or defaultValue, I can they r both missing, or even use `reset` when you have the values prepare to solve your usage problem. 

Thanks for your purpose above, but I don't think these works out for react hook form, and thank you for your passion for the lib, the lib is really a set of walk paths to get you to a solution, but obviously it has its limitation and if you are building aginst it, is probably not going to work out, and every day I have quite a few people report me with a certain use case, believe me, there is only so much I can do to help and support while maintaining it so it works for the majority. I can probably do a hot patch to make it work for your case, but it will break others. hope you find a solution or work around too.

**Reply by giovannipiller** [reply]
Ok, I understand completely.

Thank you for taking the time to reply, I really appreciated it! 🙌
